packages/vitest/src/node/hoistMocks.ts

```diff
@@ -71,58 +72,114 @@ export function hoistMocks(code: string, id: string, parse: PluginContext['parse
-
     if (node.source.value === 'vitest') {
       const code = `const ${transformImportSpecifiers(node)} = await import('vitest')\n`
       hoistedVitestImports += code
+      s.remove(node.start, node.end)
       return
     }
-    const code = transformImportDeclaration(node)
-    s.appendLeft(hoistIndex, code)
+
+    const declaration = transformImportDeclaration(node)
+    if (!declaration)
+      return null
+    s.appendLeft(hoistIndex, declaration.code)
+    return declaration.id
   }
 
   // 1. check all import statements and record id -> importName map
   for (const node of ast.body as Node[]) {
     // import foo from 'foo' --> foo -> __import_foo__.default
     // import { baz } from 'foo' --> baz -> __import_foo__.baz
     // import * as ok from 'foo' --> ok -> __import_foo__
-    if (node.type === 'ImportDeclaration')
-      hoistImport(node)
+    if (node.type === 'ImportDeclaration') {
+      const importId = defineImport(node)
+      if (!importId)
+        continue
+      s.remove(node.start, node.end)
+      for (const spec of node.specifiers) {
+        if (spec.type === 'ImportSpecifier') {
+          idToImportMap.set(
+            spec.local.name,
+            `${importId}.${spec.imported.name}`,
+          )
+        }
+        else if (spec.type === 'ImportDefaultSpecifier') {
+          idToImportMap.set(spec.local.name, `${importId}.default`)
+        }
+        else {
+          // namespace specifier
+          idToImportMap.set(spec.local.name, importId)
+        }
+      }
+    }
   }
 
+  const declaredConst = new Set<string>()
+
+  esmWalker(ast, {
+    onIdentifier(id, parent, parentStack) {
+      const grandparent = parentStack[1]
+      const binding = idToImportMap.get(id.name)
+      if (!binding)
+        return
+
+      if (isStaticProperty(parent) && parent.shorthand) {
+        // let binding used in a property shorthand
+        // { foo } -> { foo: __import_x__.foo }
+        // skip for destructuring patterns
+        if (
+          !isNodeInPattern(parent)
+            || isInDestructuringAssignment(parent, parentStack)
+        )
+          s.appendLeft(id.end, `: ${binding}`)
+      }
+      else if (
+        (parent.type === 'PropertyDefinition'
+            && grandparent?.type === 'ClassBody')
+          || (parent.type === 'ClassDeclaration' && id === parent.superClass)
+      ) {
+        if (!declaredConst.has(id.name)) {
+          declaredConst.add(id.name)
+          // locate the top-most node containing the class declaration
+          const topNode = parentStack[parentStack.length - 2]
+          s.prependRight(topNode.start, `const ${id.name} = ${binding};\n`)
+        }
+      }
+      else if (
+        // don't transform class name identifier
+        !(parent.type === 'ClassExpression' && id === parent.id)
+      ) {
+        s.update(id.start, id.end, binding)
+      }
+    },
+    onDynamicImport() {},
+    onImportMeta() {},
+  })
+
   simpleWalk(ast, {
```

Comment:
Can we include that in a previous walk for performance reasons?

Reply:
Yes, Let me make some changes in `esmWalker`